### ISSUE TITLE
WSL - vSwitch - Internal/Private - Static WSL IP/Subnet

### ISSUE THREAD

**Issue by Hyper200** [microsoft/WSL#9251]
**Is your feature request related to a problem? Please describe.**
VPN Ranges clash with randomly assigned WSL IPs, Connectivity to WSL is complicated and difficult at times, Random address make it difficult to send traffic to WSL. Docker ranges will clash occasionally.

**Describe the solution you'd like**
We would like the ability to use WSL attached to a HyperV swich where we control the subnet and the assigned IP address for vSwitch, and therefor can use init or systemd-network to assign a static ip to the WSL instance.

**Describe alternatives you've considered**
I have tried the following using a bridged adaptor, and although it works on an external switch, it does not work on an internal switch:
https://github.com/microsoft/WSL/issues/4150#issuecomment-1288152162
https://github.com/luxzg/WSL2-fixes

The following config:

[wsl2]
networkingMode=bridged
vmSwitch=WSL_external

Doesn't seem to work despite being so close to what we want. eth0 on WSL never comes up on the WSL interface.

 Could we have:

[wsl2]
networkingMode=internal
vmSwitch=WSL_internal

or

[wsl2]
networkingMode=private
vmSwitch=WSL_private

Thanks,

**Comment by luxzg** [comment]
Indeed, I tried both Private and Internal vSwitch and they seem to act as if interface is simply disconnected. While External vSwitch works. I don't see why there should be any differentiation between vSwitch types from the WSL side. If you try it with any other VM these switches work exactly the same, it's just "yet another" Ethernet adapter for the OS.

**Comment by davidshen84** [comment]
Without a static IP for the WSL instance, it is very inconvenient to use `microk8s`. I have to:

- update the kube.conf every time to use the new IP
- update `metallb` every time for the new IP range

**Comment by lm1baker** [comment]
The default range used clashes with the company network at 172.16.0.0/12
There must be a way to reconfigure it.

**Comment by Biswa96** [comment]
Newer WSL has some registry keys to configure the static IP address. Please see this discussion https://github.com/microsoft/WSL/discussions/9580

**Comment by davidshen84** [comment]
I think what we discussed in that thread is that the new WSL uses a
registry key which happens to set the IP to a static value.

But it is unclear that if that feature or that behaviour will stay the same.

**Comment by HKTY569** [comment]
> 较新的 WSL 有一些注册表项来配置静态 IP 地址。请参阅此讨论#9580

It seems that the fixed IP cannot be directly modified by modifying the registry, and the ip segment will be reset every time wsl starts

**Comment by Biswa96** [comment]
Did you change all the three registry values - NatIpAddress, NatNetwork and NatGatewayIpAddress ?

**Comment by flylan** [comment]
WSL may be a sad project, in 2025, where users are still confused about static IP, including myself. A seemingly important and basic feature has not been implemented yet. Why can't WSL directly achieve static IP configuration through external configuration like Vagrant?
